### ISSUE TITLE
Add syncing from the GitHub API to GitHub.Sync module

### ISSUE THREAD

**Issue by joshsmith** [code-corps/code-corps-api#1099]
# Problem

We should do something like the following for sync from GitHub API endpoints in `GitHub.Sync`:

```Elixir
  def issue_api(%GithubRepo{} = github_repo, %{"pull_request" => %{"url" => pull_request_url}} = payload) do
    # Pull Request
    Multi.new
    |> Multi.merge(__MODULE__, :return_repo, [github_repo])
    |> Multi.merge(__MODULE__, :fetch_pull_request, [pull_request_url])
    |> Multi.merge(GitHub.Sync.PullRequest, :sync, [payload])
    |> Multi.merge(GitHub.Sync.Issue, :sync, [payload])
    |> Multi.merge(GitHub.Sync.Comment, :sync, [payload])
    |> transact()
  end
  def issue_api(%GithubRepo{} = github_repo, %{} = issue) do
    # Issue
    Multi.new
    |> Multi.merge(__MODULE__, :return_repo, [github_repo])
    |> Multi.merge(GitHub.Sync.Issue, :sync, [issue])
    |> transact()
  end

  def comment_api(%GithubRepo{} = github_repo, %{"issue_url" => issue_url} = payload) do
    Multi.new
    |> Multi.merge(__MODULE__, :return_repo, [github_repo])
    |> Multi.merge(__MODULE__, :fetch_issue, [issue_url])
    |> Multi.merge(__MODULE__, :maybe_fetch_pull_request, [payload])
    |> Multi.merge(GitHub.Sync.PullRequest, :sync, [payload])
    |> Multi.merge(GitHub.Sync.Issue, :sync, [payload])
    |> Multi.merge(GitHub.Sync.Comment, :sync, [payload])
    |> transact()
  end

  def pull_request_api(%GithubRepo{} = github_repo, %{"issue_url" => issue_url} = payload) do
    Multi.new
    |> Multi.merge(__MODULE__, :return_repo, [github_repo])
    |> Multi.merge(__MODULE__, :fetch_issue, [issue_url])
    |> Multi.merge(GitHub.Sync.PullRequest, :sync, [payload])
    |> Multi.merge(GitHub.Sync.Issue, :sync, [payload])
    |> transact()
  end

  def return_repo(_, %GithubRepo{} = github_repo) do
    Multi.new
    |> Multi.run(:repo, fn _ -> github_repo end)
  end

  def maybe_fetch_pull_request(%{fetch_issue: %{"pull_request" => %{"url" => url}}} = changes, _) do
    fetch_pull_request(changes, url)
  end
  def maybe_fetch_pull_request(_, _), do: Multi.new

  def maybe_sync_pull_request(changes, %{} = payload) do
    Multi.new
    |> Multi.run(:sync_pull_request, fn _ -> GitHub.Sync.PullRequest.sync(changes, payload) end)
  end
  def maybe_sync_pull_request(_, _), do: Multi.new
```

**Comment by begedin** [comment]
The steps we apparently need here are

- [x] get repos for installation - This is already done as part of handling the `installation_added` event and really, needs to be done before the sync, considering our client page needs to list those repositories in order to proceed with integration and allow the user to link them to the project
  - [ ] we may need to add pagination support to the existing endpoint - #1108 
- [ ] get issues for repo 
   - this will require pagination support from #1107,
   - it will also require implementing a module in #1109 
- [ ] get comments for each issue or for repo - 
   - requires pagination support from #1107 
   - requires implementing the specific endpoint for #1110  or #1111. My vote is #1111
- [ ] once those endpoints are in, we can implement a module which receives a project and runs an issue sync, followed by a comment sync, for each github repo associated to that project

**Comment by begedin** [comment]
For convention, maybe we should consider renaming `Sync` to `WebhookSync` and add in a `ProjectSync`. Alternatively, we could stick to having everything in the same module. I'm not really sure.

**Comment by joshsmith** [comment]
Also add in:

- [ ] get pull requests for repo

This is not accounted for in "get issues".

**Comment by joshsmith** [comment]
I'm wrong above. Pull requests are returned there. However, it's frustrating that we won't have all the information for the pull request.

**Comment by joshsmith** [comment]
Closing this because everything is dealt with separately by another issue.
